### PR TITLE
Bug/sold out events not in rest api #527

### PR DESCRIPTION
**Pull request by mnelson4**
Fixes a bug which prevented sold out events from appearing in unauthenticated REST API results.
This addresses https://github.com/eventespresso/event-espresso-core/issues/527#issuecomment-397616110.
Two downsides to this fix though:
* event model queries using the model's built-in capabilities (eg over the REST API) need to do an extra join to the postmeta table
* because of that, those queries also shouldn't add the postmeta to conditions, because this is already joining to that model (and there's only ever one join to the model, which is a deficiency in the models I've long hoped to have time to address but haven't yet)

I'd still like to add unit tests

### CODE REVIEW

**Comment by mnelson4** [comment]
awaiting triage from @sethshoultes

**Comment by mnelson4** [comment]
note: this is mostly a band-aid fix (REST API still doesn't handle protected posts, and the postmeta `_previous_event_status` has always been a little hacky).
Anyways, I should discuss this band-aid fix with Brent when I get back from holiday

**Comment by mnelson4** [comment]
In Slack, Brent expressed hesitancy to adding support for multiple joins to the same table (the alternative fix for this that I was thinking of) because of performance, so he was inclined to try the band-aid fix. 
The more thorough fix would probably involve instead having a column for "previous status", or refactoring how we handle statuses on events entirely.
So I think I'll try adding some unit tests for this and see how it goes

File: core/db_models/strategies/EE_Restriction_Generator_Public.strategy.php [hunk 1]
```diff
@@ -78,4 +81,34 @@ protected function _generate_restrictions()
...
+        if ($this->model() instanceof EEM_Event) {
+            $where_conditions['OR*status'] = array(
+                'status' => $published_value,
+                'AND' => array(
+                    'Post_Meta.meta_key'   => '_previous_event_status',
+                    'Post_Meta.meta_value' => $published_value
+                )
+            );
+        } else {
+            $where_conditions['status'] = 'publish';
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply core/db_models/strategies/EE_Restriction_Generator_Public.strategy.php]: yes it should! good call!

**Comment by mnelson4** [comment]
ok, requesting another review

**Comment by mnelson4** [comment]
this needs a re-review, I forgot a changelog entry!

**Comment by mnelson4** [comment]
## Testing
* [ ] Sell out a published event. Send an UNauthenticated request to `wp-json/ee/v4.8.36/events`. It should still appear in the list
* [ ] Sell out a private or password protected event. Send an UNauthenticated request to `wp-json/ee/v4.8.36/events`. It should NOT appear in the list

**Comment by mnelson4** [comment]
Sorry @tn3rb this needs re-review since there was a conflict in the changelog.

@joshfeck would you like to test this? (the changes are unit tested, but it might not be bad to have an extra manual test). See https://github.com/eventespresso/event-espresso-core/pull/536#issuecomment-406697756 for testing notes

**Comment by joshfeck** [comment]
Tested this branch and while an un-auth'd request to /events/ does return sold out events, the sold out event's related entities aren't available. For example, in this screenshot you'll see that the sold out events do not have datetimes listed:

screen shot 2018-07-23 at 12 28 21 pm


The same thing happens on specific requests to related entities. For example, an event that's not sold out has its related datetimes listed:
https://josh.eventespresso.com/wp-json/ee/v4.8.36/events/121/datetimes

This sold out event does not have its related datetimes listed:
https://josh.eventespresso.com/wp-json/ee/v4.8.36/events/102/datetimes

**Comment by mnelson4** [comment]
oh cool catch @joshfeck. Ya this will require changing a few other things...

**Comment by mnelson4** [comment]
Thanks brent. Passing to @joshfeck for re-testing. See https://github.com/eventespresso/event-espresso-core/pull/536#issuecomment-406697756 for updated testing notes

**Comment by joshfeck** [comment]
OK the missing datetimes issue is fixed. 

Also, events that were previously set to Private post status before selling out are not included in the event list via an unauth'd GET request to /events.

However, sold out events that were previously set to Password Protected are included in the event lis list with an unauth'd GET request to /events.
e.g.
https://josh.eventespresso.com/wp-json/ee/v4.8.36/events?include=Datetime.DTT_EVT_start,EVT_ID,EVT_name&sort=DESC&order_by=EVT_ID

**Comment by joshfeck** [comment]
FWIW, Password Protected events are visible on unauth'd GET requests to the /events/ route **on master**. So this might be a separate issue.

**Comment by joshfeck** [comment]
OK the missing datetimes issue is fixed.

Also, events that were previously set to Private post status before selling out are not included in the event list via an unauth'd GET request to /events.

**Pull request merged**